### PR TITLE
Added select2 loading all tag, Ajax tag list generation and single tag search Ajax

### PR DESCRIPTION
**Pull request by coolRaider**
**What this PR does:**
- Integrated Select2
- Added Ajax tag list loading
- Added Ajax search results population for a single tag search item

### CODE REVIEW

File: app/assets/javascripts/articles.coffee [hunk 1]
```diff
@@ -1,5 +1,41 @@
...
+
+sel = ->
+  $("#tag-select").select2
+    theme: "bootstrap"
+    allowClear: true
+    placeholder: "type in the tags here"
+    ajax:
+      url: "/tags"
+      dataType: 'json'
+      delay: 250
```
Comment (joney000) [inline app/assets/javascripts/articles.coffee:13]: is delay time too less , as no of tags are increased in databases ?

File: app/controllers/tags_controller.rb [hunk 2]
```diff
@@ -1,10 +1,25 @@
...
+      format.js {}
+      format.json {render json: @tags}
+    end
   end
 
   def show
+    @q = Article.ransack(params[:q])
     @tag = ActsAsTaggableOn::Tag.find(params[:id])
-    @articles = Article.tagged_with(@tag.name).paginate(:page => params[:page], :per_page => 4)
+    @articles = @q.result(distinct: true).tagged_with(@tag.name).paginate(:page => params[:page], :per_page => 1)
```
Comment (joney000) [inline app/controllers/tags_controller.rb:18]: per page only one blog ?

File: app/views/tags/show.js.haml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+:plain
+  $('#results').html("#{ j (render partial: 'articles/list', locals: {articles: @articles}) }")
+  $('.pagination a').attr('data-remote', 'true');
```
Comment (joney000) [inline app/views/tags/show.js.haml:3]: new line at EOF.

File: app/assets/javascripts/articles.coffee (hunk 1 above)
Comment (coolRaider) [inline app/assets/javascripts/articles.coffee:13]: no, if we keep it too less, then the application would appear laggy.

File: app/controllers/tags_controller.rb (hunk 2 above)
Comment (coolRaider) [inline app/controllers/tags_controller.rb:18]: this is done for testing purposes, to check if pagination is working as expected.

**Pull request merged**